### PR TITLE
Preparing for functional tests

### PR DESCRIPTION
**Pull request by moozzyk**
Early feedback solicited @anurse @BrennanConroy 

Making test infra (deployer and fixtures) shareable across test projects. Added a server and a project for functional tests. Added a single functional test to prove the changes work. 

For now this is C#/net451 - I am not sure how/if we can use this to run JS functional tests - will investigate.

### CODE REVIEW

File: test/Microsoft.AspNetCore.SignalR.FunctionalTests/project.json [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  "dependencies": {
+    "Microsoft.AspNetCore.Server.IntegrationTesting": "0.2.0-*",
+    "xunit": "2.2.0-*",
+    "dotnet-test-xunit": "2.2.0-*",
+    "Microsoft.Extensions.PlatformAbstractions": "1.1.0-*",
+    "Microsoft.Extensions.Logging": "1.1.0-*",
+    "Microsoft.Extensions.Logging.Console": "1.1.0-*",
+    "Microsoft.Extensions.RuntimeEnvironment.Sources": "1.1.0-*",
+    "Microsoft.AspNetCore.SignalR.Testing.Common": "1.1.0-*",
+    "Microsoft.AspNet.SignalR.Client": "2.2.1"
```
Comment (BrennanConroy) [inline test/Microsoft.AspNetCore.SignalR.FunctionalTests/project.json]: Remove this one

Comment (moozzyk) [inline test/Microsoft.AspNetCore.SignalR.FunctionalTests/project.json]: I will.

File: test/Microsoft.AspNetCore.SignalR.FunctionalTests.Server/Hubs/EchoHub.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
+﻿
```
Comment (BrennanConroy) [inline test/Microsoft.AspNetCore.SignalR.FunctionalTests.Server/Hubs/EchoHub.cs]: remove line

File: test/Microsoft.AspNetCore.SignalR.FunctionalTests/HubTests.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
+﻿
```
Comment (BrennanConroy) [inline test/Microsoft.AspNetCore.SignalR.FunctionalTests/HubTests.cs]: remove line

**Comment by BrennanConroy** [comment]
Looks good :shipit:

**Comment by anurse** [comment]
I like it. Though I do wonder if maybe we should go back to @BrennanConroy 's suggestion of just having one test server. I hesitated because I didn't want it to become a dumping ground but we don't really need a huge number of hubs. The Compat Tests were not as complex as I had originally feared, so they really only have the one hub and the one PC.

**Comment by moozzyk** [comment]
I don't see a problem with just one server - especially that for basic functional tests we will probably use the same chathub. I would like to keep parts of the code that enable multiple servers though - just in case we need more servers in the future. I would also remove "Compat" from the name if we end up having just one server.

**Comment by anurse** [comment]
Agreed on keeping the multi-server parts. We don't want to restrict ourselves in case we find we need multiple server projects in the future

**Comment by anurse** [comment]
I'd say let's just merge the servers. We're also going to have JavaScript-based compat tests and functional tests which can live in the same server anyway. My PR #223 starts on some infrastructure to the CompatTest server to support Jasmine-based tests written in TypeScript so we can unify that into one server (just with different directories, and possibly even different base HTML pages, for the compat and functional tests)

**Comment by moozzyk** [comment]
:up::date:

File: test/Microsoft.AspNetCore.SignalR.FunctionalTests/project.json [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    "xunit": "2.2.0-*",
+    "dotnet-test-xunit": "2.2.0-*",
+    "Microsoft.Extensions.PlatformAbstractions": "1.1.0-*",
+    "Microsoft.Extensions.Logging": "1.1.0-*",
+    "Microsoft.Extensions.Logging.Console": "1.1.0-*",
+    "Microsoft.Extensions.RuntimeEnvironment.Sources": "1.1.0-*",
+    "Microsoft.AspNetCore.SignalR.Testing.Common": "1.1.0-*"
+  },
+
+  "frameworks": {
```
Comment (BrennanConroy) [inline test/Microsoft.AspNetCore.SignalR.FunctionalTests/project.json]: Shouldn't this have netcoreapp?

Comment (BrennanConroy) [inline test/Microsoft.AspNetCore.SignalR.FunctionalTests/project.json]: Oh wait... we don't have a client yet

File: test/Microsoft.AspNetCore.SignalR.FunctionalTests/project.json [hunk 5]
```diff
@@ -0,0 +1,27 @@
+﻿{
+  "version": "1.0.0-*",
+  "buildOptions": {
+    "allowUnsafe": true,
+    "keyFile": "../../tools/Key.snk"
+  },
+  "testRunner": "xunit",
+
+  "dependencies": {
+    "Microsoft.AspNetCore.Server.IntegrationTesting": "0.2.0-*",
```
Comment (BrennanConroy) [inline test/Microsoft.AspNetCore.SignalR.FunctionalTests/project.json]: nit: sort the dependencies

**Pull request merged**